src/QMCWaveFunctions/BsplineFactory/SplineR2RAdoptor.h

```diff
@@ -222,26 +223,54 @@ struct SplineR2RSoA: public SplineAdoptorBase<ST,3>
     const PointType& r=P.activeR(iat);
     PointType ru;
     int bc_sign=convertPos(r,ru);
-    spline2::evaluate3d(SplineInst->spline_m,ru,myV);
-    assign_v(bc_sign,myV,psi);
+
+    #pragma omp parallel
+    {
+      int first, last;
+      FairDivideAligned(myV.size(), getAlignment<ST>(),
+                        omp_get_num_threads(),
+                        omp_get_thread_num(),
+                        first, last);
+
+      spline2::evaluate3d(SplineInst->spline_m,ru,myV,first,last);
+      assign_v(bc_sign,myV,psi,first,last);
+    }
   }
 
   template<typename VM, typename VAV>
   inline void evaluateValues(const VirtualParticleSet& VP, VM& psiM, VAV& SPOMem)
   {
-    const size_t m=psiM.cols();
-    for(int iat=0; iat<VP.getTotalNum(); ++iat)
+    #pragma omp parallel
     {
-      Vector<TT> psi(psiM[iat],m);
-      evaluate_v(VP,iat,psi);
+      int first, last;
+      FairDivideAligned(myV.size(), getAlignment<ST>(),
+                        omp_get_num_threads(),
+                        omp_get_thread_num(),
+                        first, last);
+
+      const size_t m=psiM.cols();
+      for(int iat=0; iat<VP.getTotalNum(); ++iat)
+      {
+        const PointType& r=VP.activeR(iat);
+        PointType ru;
+        int bc_sign=convertPos(r,ru);
+        Vector<TT> psi(psiM[iat],m);
+
+        spline2::evaluate3d(SplineInst->spline_m,ru,myV,first,last);
+        assign_v(bc_sign,myV,psi,first,last);
+      }
     }
   }
 
   inline size_t estimateMemory(const int nP) { return 0; }
 
   template<typename VV, typename GV>
-  inline void assign_vgl(int bc_sign, VV& psi, GV& dpsi, VV& d2psi)
+  inline void assign_vgl(int bc_sign, VV& psi, GV& dpsi, VV& d2psi, int first = 0, int last = -1) const
   {
+    // protect last
+    last = last<0 ? kPoints.size() : (last>kPoints.size() ? kPoints.size() : last);
+
+    const ST cone = (bc_sign &1)? -1:1;
```

Comment:
Why cone?

Reply:
Do you have suggestion for a better name? cone means constant one, maybe I should put one_with_sign?